File: services/scanner-function/path_parser.py
import re
from typing import Optional

# Standard render path: projects/{PROJECT}/{SEQUENCE}/{SHOT}/render/{VERSION}/{filename}
# With episode:         projects/{PROJECT}/{EPISODE}/{SEQUENCE}/{SHOT}/render/{VERSION}/{filename}
RENDER_PATTERN = re.compile(
    r"^projects/"
    r"(?P<project>[^/]+)/"
    r"(?:(?P<episode>EP\w+)/)?"   # optional episode segment (EP01, EP02, ...)
    r"(?P<sequence>[^/]+)/"
    r"(?P<shot>[^/]+)/"
    r"render/"
    r"(?P<version>v\d+[^/]*)/"   # v001, v002_colorfix, etc.
    r"(?P<filename>[^/]+)$"
)

# .ready is the sentinel extension for render sequences.
# When a renderer finishes writing all frames to the VAST view it drops a
# zero-byte "<shot>_<version>.ready" file in the same render directory.
# The DataEngine ElementCreated trigger fires on this file, and the scanner
# aggregates the whole frame sequence into a single asset instead of
# ingesting every individual EXR frame.
SENTINEL_EXTENSION = ".ready"

SUPPORTED_EXTENSIONS = {".exr", ".mov", ".dpx", ".audio", ".wav", ".vdb", ".usd", ".usda", ".usdc", ".usdz", ".abc"}


def parse_render_path(key: str) -> Optional[dict]:
    """
    Parse an S3 object key into structured VFX hierarchy fields.
    Returns None if the key does not match the expected render path pattern
    or the file extension is not supported.

    When the key ends with SENTINEL_EXTENSION (.ready), the sentinel is
    detected and the result includes ``"is_sentinel": True`` along with the
    render directory path so the caller can aggregate the frame sequence.
    The ``filename`` in a sentinel result is the directory path, not the
    .ready file itself, so the asset title reflects the sequence, not the
    trigger file.
    """
    m = RENDER_PATTERN.match(key)
    if not m:
        return None
    filename = m.group("filename")
    ext = "." + filename.rsplit(".", 1)[-1].lower() if "." in filename else ""

    # Sentinel path: aggregate the whole render directory as one asset.
    if ext == SENTINEL_EXTENSION:
        # The render directory is the parent of the .ready file.
        render_dir = key[: key.rfind("/") + 1]  # e.g. "projects/PROJ/SEQ/SH/render/v001/"
        # Use the directory path (without trailing slash) as the representative asset path.
        asset_path = render_dir.rstrip("/")
        return {
            "project_code":  m.group("project"),
            "episode_code":  m.group("episode"),
            "sequence_code": m.group("sequence"),
            "shot_code":     m.group("shot"),
            "version_label": m.group("version"),
            "filename":      asset_path,   # directory path — used as asset title/source URI
            "extension":     SENTINEL_EXTENSION,
            "is_sentinel":   True,
        }

    if ext not in SUPPORTED_EXTENSIONS:
        return None
    return {
        "project_code":  m.group("project"),
        "episode_code":  m.group("episode"),   # None if not episodic
        "sequence_code": m.group("sequence"),
        "shot_code":     m.group("shot"),
        "version_label": m.group("version"),
        "filename":      filename,
        "extension":     ext,
        "is_sentinel":   False,
    }

File: services/control-plane/src/scanner/path-parser.ts
/**
 * S3 render-path parser — TypeScript port of the legacy Python
 * scanner-function `path_parser.py`.
 *
 * Pattern accepted (with optional episode segment for episodic shows):
 *   projects/{PROJECT}/{SEQUENCE}/{SHOT}/render/{VERSION}/{filename}
 *   projects/{PROJECT}/{EPISODE}/{SEQUENCE}/{SHOT}/render/{VERSION}/{filename}
 *
 * Sentinel handling: keys ending in `.ready` are treated as a marker that
 * the renderer finished writing every frame in the directory. The parser
 * returns `is_sentinel: true` and `filename` is set to the render directory
 * (without trailing slash) so the caller ingests the whole sequence as a
 * single asset rather than every individual EXR.
 *
 * Pure module — no I/O, no globals. Behavior must match
 * services/scanner-function/path_parser.py until that file is removed.
 */

export const SENTINEL_EXTENSION = ".ready";

export const SUPPORTED_EXTENSIONS: ReadonlySet<string> = new Set([
  ".exr", ".mov", ".dpx", ".audio", ".wav", ".vdb",
  ".usd", ".usda", ".usdc", ".usdz", ".abc",
]);

const RENDER_PATTERN =
  /^projects\/(?<project>[^/]+)\/(?:(?<episode>EP\w+)\/)?(?<sequence>[^/]+)\/(?<shot>[^/]+)\/render\/(?<version>v\d+[^/]*)\/(?<filename>[^/]+)$/;

export interface ParsedRenderPath {
  projectCode: string;
  episodeCode: string | null;
  sequenceCode: string;
  shotCode: string;
  versionLabel: string;
  filename: string;
  extension: string;
  isSentinel: boolean;
}

export function parseRenderPath(key: string): ParsedRenderPath | null {
  const match = RENDER_PATTERN.exec(key);
  if (!match || !match.groups) return null;

  const filename = match.groups.filename;
  const dotIdx = filename.lastIndexOf(".");
  const extension = dotIdx === -1 ? "" : "." + filename.slice(dotIdx + 1).toLowerCase();

  if (extension === SENTINEL_EXTENSION) {
    // Sentinel — represent the whole render directory as the "filename".
    const slashIdx = key.lastIndexOf("/");
    const renderDir = slashIdx === -1 ? key : key.slice(0, slashIdx + 1);
    return {
      projectCode: match.groups.project,
      episodeCode: match.groups.episode ?? null,
      sequenceCode: match.groups.sequence,
      shotCode: match.groups.shot,
      versionLabel: match.groups.version,
      filename: renderDir.replace(/\/$/, ""),
      extension: SENTINEL_EXTENSION,
      isSentinel: true,
    };
  }

  if (!SUPPORTED_EXTENSIONS.has(extension)) return null;

  return {
    projectCode: match.groups.project,
    episodeCode: match.groups.episode ?? null,
    sequenceCode: match.groups.sequence,
    shotCode: match.groups.shot,
    versionLabel: match.groups.version,
    filename,
    extension,
    isSentinel: false,
  };
}
